### ISSUE TITLE
Payment Gateway Tokenization API

### ISSUE THREAD

**Issue by mikejolley** [woothemes/woocommerce#9095]
Allow gateways to declare support for tokenization and offer standard functions to add, remove and retrieve tokens for customers. Any gateway that does this will work with Preorders, Deposits, & and any future extensions that need tokens (theoretically).

Some notes: https://trello.com/c/nI3r1dbe/70-tokenised-payment-api

---

## What this API covers

This API is for **storing** and **managing** tokens for gateways and customers. The following functionality will be covered.

1. Validating and storing a token for a gateway, and associating said token with an **order** or **customer**
2. Methods for token retrieval
3. Customer facing UI to view their saved tokens (cards) and add/remove them (see SkyVerge's UI https://www.skyverge.com/wp-content/uploads/2015/07/woocommerce-authorize-net-cim-my-payments.png)

## Token Class

The token class represents a single token stored in the database. Each token can have the follow data:

- gateway ID
- token ID
- token
- last 4
- expiry date
- card type
- "default"

### Methods

- WC_Token::get_id()
- WC_Token::get_last4()
- WC_Token::get_token()
- WC_Token::get_expiry_date()
- WC_Token::get_card_type()
- WC_Token::get_customer_id() - user ID or false if not associated with one
- WC_Token::get_gateway_id()
- WC_Token::is_default()
- WC_Token::set_* for the above

### Notes

- SkyVerge's implementation has an 'image' for each token. I think images for card types should be separate as it can be used elsewhere in the UI.
- "Credit card" type I would expect to be the most popular so other types (bank, cheque?) could extend this, but I am not certain if there would be other common use cases. @allendav and @maxrice could probably give feedback there.
- I think this makes sense to be in a separate custom database table to keep it separate and secure.
- Orders can be linked with tokens by ID (think about guest orders). Could be _wc_tokens meta at order level - serialised array of ids.

## WC_Tokens Class

Controller and factory for tokens in general.

- [ ] WC_Tokens::create( $args ) ?
- [ ] WC_Tokens::delete( $token_id ) ?
- [ ] WC_Tokens::get_customer_tokens( $user_id )
- [ ] WC_Tokens::get_order_tokens( $order_id ) - will return tokens attached to an order, and an orders customer if applicable

## Hooks/Actions
- token_added
- token_updated
- token_deleted
- token_expired

---

The above is a draft and pending feedback.

**Comment by mikejolley** [comment]
@allendav @maxrice @thenbrent @claudiosmweb started a mort in depth draft for this. Please give your feedback.

**Comment by thenbrent** [comment]
Looks solid to me.

Something Subscriptions could use would be a `WC_Tokens::validate()` method. This could also be bundled into `WC_Tokens::create()` as long as there is a way to catch/handle the error. Subscriptions needs this because it allows store managers to [manually enter tokens](https://docs.woothemes.com/document/subscriptions/add-or-modify-a-subscription/#section-8).

> The token class represents a single token stored in the database.

How would this be used for gateways which have multiple pieces of data for each payment token? e.g. Stripe has a customer and card token.

Would you store both as separate tokens and associate the card token with the customer token somehow?

If so, that probably means there needs to be a way to link tokens or associate them with others too. I'm sure @maxrice will have a better idea around this.

**Comment by greguly** [comment]
+1 for WC_Tokens::validate(), it would be great if validate() called the gateway to check the token

**Comment by maxrice** [comment]
Excited to have this going in core! My initial thoughts:

* I'd vote for a nomenclature change, `WC_Payment_Token` over just plain `WC_Token`. "Tokens" as a general word can mean a lot of different things when taken out of the payment context and I think it benefits us to be specific.

* :thumbsup: for associating a token with an order and/or customer. We've opted to always associate the token with a WP user (in user meta), except for Pre-Orders where the token is saved to order meta for the later pre-order transaction.

* I'd try to keep the token class somewhat abstracted from the concept of a credit card and instead treat it like a generic payment method. For example, a credit card has both an expiration date, card type, and last 4 but an eCheck would only have a last 4. For a PayPal account (the new Braintree integration can save these), none of that data would be present. An abstract class with sub-classes for the big 2 (credit card/eCheck) would be ideal, provided those are easily overridable (and individual data is filterable) via the factory. I think this could be done by calling a method on the token's payment gateway to return the token object, so implementing gateways could override that method as needed.

* One of the big challenges we ran into initially with our framework token handling was keeping local tokens in sync with remote tokens, e.g. a customer adds a payment method locally on the site and then an admin later changes something about the token in the gateway's control panel, like updating the expiration date. The customer would see the wrong expiration date when they checked out next. However, you don't want to _always_ retrieve the tokens over the API as that's slow and it doesn't account for situations when there might be an API issue that's just affecting the tokens.

   The end result was that a gateway could declare support for retrieving remote tokens, which the framework would then use in lieu of the local tokens. We provided a way for gateways to merge local token data into the remote token data, because some gateways (see Authorize.net CIM) provide _less_ information when requesting the tokens over their API than what you get when you save the token (e.g. card type is not returned). To keep page loads snappy we cache the tokens in a 1 minute transient so checkout updates aren't affected by the remote API calls. TL, DR: make it easy to support overriding the local data with token data retrieved via an API call

* We opted to have an image associated for each token as it's convenient when rendering both the "My Payment Methods" UI in the my account area and the saved payment methods available in the checkout. Even if the images are abstracted away from the actual token class, I think you'd still want a method like `$token->get_image_url()`. Some gateways even provide this info in the token response (see [Braintree](https://developers.braintreepayments.com/reference/response/credit-card/php#image_url))

* I don't have a strong opinion on where token data should be stored, but if it's a custom table I'd definitely want the ability to store custom data against the token, either in a separate token meta table, or in a meta column (assume it'd be serialized in that case)

* I think you'd want a separate method on the orders class to return tokens associated with it, rather than a method on the `WC_Tokens` class, e.g. `$order->get_payment_tokens()`

* I think an `WC_Tokens::update( $token_id, $data )` method would be convenient to have.

What @thenbrent asked about separate customer/card tokens is a good question. We've found that some gateways require both a customer ID and token ID when creating a transaction from a saved token. Others simply require the token ID and no other info. So, I think the best approach for core is to have a customer ID as an optional attribute for the token along with associated get/set methods. For gateways that don't require it, it can simply return `null`. This should neatly handle gateways like Stripe, and if other gateways require even more information, they can make use of token meta (like I mentioned above).

Are there plans to add a method to render [saved payment methods](https://docs.woothemes.com/wp-content/uploads/2013/04/woocommerce-authorize-net-saved-card-checkout.png) at checkout as part of this?

One last thought is that I'd lean towards making this whole implementation as flexible as possible. The one big thing we've learned from working with 25+ gateway integrations is that they all do things slightly differently, especially when dealing with tokenization. Having lots of actions & filters and the ability to override methods or entire classes will be key to driving adoption of this core feature in extensions.

For reference, here's the parts in our payment gateway framework that are related to this and might be useful for inspiration/ideas:
* [Gateway tokenization methods](https://github.com/skyverge/wc-plugin-framework/blob/wc-plugin-framework-4.1.2/woocommerce/payment-gateway/class-sv-wc-payment-gateway-direct.php#L1087-L1810)
* [Payment Token class](https://github.com/skyverge/wc-plugin-framework/blob/wc-plugin-framework-4.1.2/woocommerce/payment-gateway/class-sv-wc-payment-gateway-payment-token.php)

**Comment by mikejolley** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;nI3r1dbe&#x2F;70-tokenised-payment-api"><img src="https:&#x2F;&#x2F;github.trello.services&#x2F;images&#x2F;trello-icon.png" width="12" height="12"> Tokenised Payment API</a>

**Comment by allendav** [comment]
I like to work from use cases - I think what's proposed above will work well, for these use cases I've identified:

#1:  A customer puts a non-subscription item in their cart and proceeds to checkout.  In checkout, they sign up for a new account with XYZ gateway.  XYZ gateway returns a customer object (an ID) and a payment object (token) that can be used for future payments to the merchant’s site.  The customer comes back later and is able to order something else using the saved payment.  (The XYZ gateway requires both the customer object and the payment object.)

#2: A customer that has previously ordered from a merchant is worried their credit card may have been stolen.  They log into the merchant site and delete that card from their saved payment options.

#3: A customer that has previously ordered from a merchant calls the merchant to inquire about an out-of-stock item.  The merchant realizes they have one left and so they create the order and charge the customer using the customer’s saved info.  (The XYZ gateway is able to fetch and use both the customer object and the payment object.)

#4: A customer has a subscription and their credit card is expiring next month (i.e. the payment will fail.)  An automatic email is generated to the customer asking them to provide a new payment method.  The customer logs in and adds a new card and tells the merchant’s site to use it for future recurring payments.  The site automatically begins using that payment method for their subscription.

#5: The customer has multiple payment methods saved (e.g. credit cards at two different banks) and instead of remembering the last four digits is offered the ability to set a “friendly name” for each payment method, e.g. “US Bank Credit Card”

#6: The merchant has a catastrophic failure of their site.  They had used VaultPress to backup their site before the failure and they are able to restore from backup successfully, including customer objects and payment objects.  Subscription revenue continues to pour in, uninterrupted.

#7A: The customer has a subscription and wants to switch it to a new credit card serviced by the same XYZ gateway.  They are able to log in, review their account, add the card, and switch the subscription to the new card.

#7B: The customer has a subscription and wants to switch it to a payment gateway ABC.  They are able to log in, review their account, add the payment method, and switch the subscription to the new payment method.

I guess my only concern is if we put this in a separate table, will backup solutions like "VaultPress" be able to find it?

An then, for composition, I completely agree with @maxrice on the need for flexibility.  I would nail down a small handful of field names for these objects (e.g. token, token_expires, etc) but make adding methods + meta easy to do, e.g. (I was sorely tempted to fire up a UML drawing tool):

A payment gateway will declare whether or not it “supports” the Payment Gateway Tokenization (PGT) API.

PGT enabled gateways will support creating, reading, updating, and deleting customer objects.
(The ROW ID of the customer objects will be saved in a gateway based key on usermeta, e.g. {$gateway}_customer_IDs)
A customer object is not required.  A user may have 0 or more customer object references in their usermeta.
A customer object may optionally have a ‘token’ and ‘token_expires’ date time field if the gateway supports the concept of a customer token.
A customer object may optionally have a ‘last_validated’ date time field if the gateway supports validating the token (i.e. to avoid having to issue multiple requests to a gateway in short order)
If the token never expires, token_expires is set to false.

PGT enabled gateways will support creating, reading, updating, and deleting payment method objects.
(The ROW ID of the payment method objects will be saved in a gateway based key on usermeta, e.g. {$gateway}_payment_method_IDs.)
A payment method object is not required.  A user may have 0 or more payment method object references in their usermeta.
A payment method object may optionally have a ‘token’ and ‘token_expires’ date time field if the gateway supports the concept of a (reusable) payment method token.
If the token never expires, token_expires is set to false.
A payment method object may optionally have a ‘last_validated’ date time field if the gateway supports validating the token (i.e. to avoid having to issue multiple requests to a gateway in short order)
If a transaction also requires passing data from a customer object, the payment method object will include a reference to the customer object ID e.g. ’custID’ (since there possibly could be more than one customer object for a given customer for a given gateway)
A payment method object will include an optional ‘friendly_name’ field
A payment method object will include an ‘method’ field with values of ‘credit_card’, etc.
The supported list of method values will be filterable
A payment method object will include a meta field (i.e. to store method specific details)

**Comment by justinshreve** [comment]
tl;dr This is a first pass at a Payment Token API taking the original GH proposal, provided feedback, and the Trello card into consideration. The first pass only focuses on validating, storing, and retrieving (no UI), it can still use some cleanup, and there are no committed tests yet. There are other improvements to make (validation for example can check to make sure expiry is correct format, etc) but I wanted to go ahead and post.

You can see the first pass here: https://github.com/woothemes/woocommerce/compare/master...justinshreve:token-api

I went with the approach recommended by @maxrice to try to make this extendable and not tied down to just cards:

* There is an abstract WC_Payment_Token class. It contains fields shared by all payment tokens: The token itself, token ID, gateway ID, customer ID if there is one, default status, and 'type'. This data is stored in a dedicated tokens table.
* There are two bundled payment token types: Credit Card & eCheck.
    * Credit cards contain expiration fields, last 4, and card type (visa, master card, etc).
    * eCheck just contains last 4.
    * Other extensions can add new token types.
 * Meta is supported so extra fields can be stored if new payment token types are added. Anything outside of the above shared fields is stored in a meta table, but easily exposed. 

I want to work on getting some good committable tests written next, improving upon the above code, and then finally UI stuff. I imagine since we can support different token types (which might display differently as a row on checkout/my account) we'll need some fallback UI plus a way for token types (card, eCheck, ..) to say what they are displaying.

**Comment by mikejolley** [comment]
Closing in favour of #10260
